[PXCT-1779] brick docs rework (#2944)
arduino/docs-content: content/software/app-lab/5.bricks/4.bricks-reference/bricks-reference.md

---
title: Bricks Architecture and Configuration Reference
description: A technical reference for Bricks inside Arduino App Lab, detailing thread lifecycle, YAML schemas, and configuration specifications.
overwriteSidebar: Reference
tags:
  - Bricks
  - Architecture
  - Python
  - Docker
  - Configuration
---

This document details the underlying execution model, thread lifecycle management, file directory structures, and configuration specifications for both preinstalled official bricks and custom bricks in the Arduino App Lab ecosystem.

## Core Concepts & Lifecycle Management

Bricks follow one of two execution patterns inside the `arduino.app_utils` framework, depending on whether you implement them as plain functions or managed classes.

> 

Reviewer: ```suggestion

<Alert type="warning">**Warning:** Because `App.run()` blocks the main thread, you must perform all your brick initialization, variable reading, and setup logic *before* calling `App.run()`.</Alert>

```

Preserve the sole relevant information from the otherwise off-topic "App.run() Thread Management" section.

## App.run() Thread Management

By default, the `App.run()` function blocks the main thread. It initiates an infinite sleep loop to keep the process alive and ensure that background brick threads continue executing.

### The `user_loop` Parameter

If you pass a callable function to `App.run(user_loop=...)`, the orchestrator will execute that function repeatedly inside the infinite loop on the main thread. This is ideal for running continuous, repetitive code in simple applications without manually configuring dedicated background threads.

```py
from arduino.app_utils import App
import time

def my_continuous_loop():
    # This runs repeatedly on the main thread
    print("Main thread active...")
    time.sleep(1)

# App.run blocks the main thread and repeatedly executes my_continuous_loop
App.run(user_loop=my_continuous_loop)
```

<Alert type="warning">**Warning:** Because `App.run()` blocks the main thread, you must perform all your brick initialization, variable reading, and setup logic *before* calling `App.run()`.</Alert>

> 

Reviewer: ```suggestion
```

This information is completely off topic for the Brick reference.

## Custom Bricks Directory Structure

Custom bricks are stored locally in your application's `bricks/` folder. The folder name must match the custom brick's configuration ID.

```text
my-app/
├── app.yaml
├── python/main.py
└── bricks/

>     └── my_custom_brick/            # The Brick ID (must match folder name)

Reviewer: ```suggestion
    └── my_custom_brick/            # The Brick ID (must match id key in manifest)
```

Previously, this would be read as the equivalent of saying "the folder name must match the folder name".

  - **Required Variables:** If `default_value` is missing, the user must provide a value in the UI before running the app.
  - **Secrets:** Setting `secret: true` will mask the input field in the App Lab UI.
- **`requires_display`**: When set to `webview`, this indicates that the brick exposes a graphical interface. This signals the App Lab desktop application to automatically open a new window or browser tab directed to the brick's web interface as soon as the app finishes starting.
- **`requires_services`**: List of core system services (IDs) that must be running for this brick to function (e.g., `arduino:genie` for LLM bricks).
- **`required_devices`**: Defines the *classes* of hardware devices the brick requires (e.g., `camera`, `microphone`, `speaker`). When the app is started, the `arduino-app-cli` performs startup validation to ensure at least one physical device of that class is currently connected to the board. If not, an error is raised.
- **`mount_devices_into_container`**: When `true`, hardware devices (like `/dev/video0`) requested in `required_devices` are automatically mounted into the brick's Docker container.

## Application Configuration (`app.yaml`)

> Bricks must be declared in the `bricks` list within the application's `app.yaml` descriptor file. The CLI parses this file to configure the orchestrator.

Reviewer: ```suggestion
Bricks must be declared in the `bricks` list within the application's `app.yaml` descriptor file.
```

The removed sentence doesn't provide anything of value to the reader.

Arduino App Lab automatically manages this section when using the visual interface.

```yaml
name: "My App Name"
description: "A description of the app"

> icon: "😀"                 # Must be a single valid emoji

Reviewer: ```suggestion
icon: "😀"                 # Must be a single emoji
```

Unless there is a specific restriction on which emoji may be used, any emoji will be "valid". If there is a restriction (which I doubt), then the specifics must be documented.

Author: @per1234 arduino-app-cli [validates the emoji here](https://github.com/swikstrom/arduino-app-cli/blob/babddae79207618e856adb89d957b4cb4623ecfc/internal/orchestrator/app/parser.go#L130-L143), using the [go.emoji](https://github.com/Andrew-M-C/go.emoji) library. That library appears to reliably parse emoji modifiers (like [skin tone](https://github.com/Andrew-M-C/go.emoji/blob/master/internal/official/emoji-sequences.txt#L1022-L1026)) correctly. In conclusion, I think there are no known restrictions 👍

ports: []                 # Array of exposed ports for the app
bricks:
  # Official brick (simple declaration)
  - arduino:web_ui: {}

  # Custom brick with configuration variables
  - my_custom_brick:
      variables:
        NAME: "Arduino"
      model: "my_custom_model_id"  # Optional: Bind a specific AI model
```

## Containerization and Networking Constraints

> Bricks can execute within an isolated background Docker container. You configure and manage this behavior using the following reference-grade rules:

Reviewer: ```suggestion
Bricks can execute within an isolated background Docker container. You can configure and manage this behavior:
```

Remove word salad.

- **Service Dependencies (`requires_services`):** If your brick relies on core containerized system services (such as local AI inference engines), list their system IDs here (e.g., `arduino:genie`). The orchestrator ensures these system services are active before launching your brick.

> - **Hardware Access (`mount_devices_into_container`):** Setting this boolean field to `true` instructs the orchestrator to automatically mount any physical hardware devices matching the classes in `required_devices` (e.g., `/dev/video0`) into the brick's Docker container.

Reviewer: ```suggestion
- **Service Dependencies:** If your brick relies on core containerized system services (such as local AI inference engines), list their system IDs (e.g., `arduino:genie`) in the `requires_services` key of `brick_config.yaml`. The orchestrator ensures these system services are active before launching your brick.
- **Hardware Access:** Setting the `mount_devices_into_container` key of `brick_config.yaml` to `true` instructs the orchestrator to automatically mount any physical hardware devices matching the classes in `required_devices` (e.g., `/dev/video0`) into the brick's Docker container.
```

Specify that these are referring to `brick_config.yaml`.

- **Virtual Compose Network:** When defining custom services inside a standard [Docker Compose file](https://docs.docker.com/reference/compose-file/) named `brick_compose.yaml`, the orchestrator runs those containers inside an isolated virtual network. The custom brick's Python code running inside the main application container communicates with these services via network APIs (HTTP, WebSockets, or TCP/IP) using the compose service name as the hostname.

> - **Pre-compiled Images:** Arduino App Lab doesn't support compiling custom Docker images directly on the board from a `Dockerfile` via the `build` directive in `brick_compose.yaml`. You must build custom images beforehand, publish them to a public registry (such as GitHub Container Registry or Docker Hub), or load them onto the board manually using `docker load`.

Reviewer: ```suggestion
- **Custom Images:** Arduino App Lab doesn't support building custom Docker images directly on the board from a `Dockerfile` via the `build` directive in `brick_compose.yaml`. You must build custom images beforehand, publish them to a public registry (such as GitHub Container Registry or Docker Hub), or load them onto the board manually using `docker load`.
```

Correct terminology.

## Official Repositories

| Repository | Purpose |
| :---- | :---- |
| `arduino/app-bricks-py` | Source code for preinstalled official bricks. |
| `arduino/app-bricks-examples` | Reference applications demonstrating brick integration. These are accessible in Arduino App Lab. |
| `arduino/arduino-app-lab` | The Arduino App Lab application. |

> | `arduino/arduino-app-cli` | Command-line tool for managing and validating apps. |

Reviewer: ```suggestion
| [`arduino/app-bricks-py`](https://github.com/arduino/app-bricks-py) | Source code for preinstalled official bricks. |
| [`arduino/app-bricks-examples`](https://github.com/arduino/app-bricks-examples) | Reference applications demonstrating brick integration. These are accessible in Arduino App Lab. |
| [`arduino/arduino-app-lab`](https://github.com/arduino/arduino-app-lab) | The Arduino App Lab application. |
| [`arduino/arduino-app-cli`](https://github.com/arduino/arduino-app-cli) | Command-line tool for managing and validating apps. |
```
